server/app/initializers/01_symmetric_encryption.rb

```diff
@@ -1,10 +1,38 @@
 require 'symmetric-encryption'
 require 'symmetric_encryption/extensions/mongoid/encrypted'
+require 'hkdf'
 
-if ENV['VAULT_KEY'] && ENV['VAULT_IV']
+if vault_key = ENV['VAULT_KEY']
+  # generate strong keys using entropy derived from the weak (human-readable) input key
+  hkdf = HKDF.new(vault_key)
+
+  # primary cipher used for encryption
+  # not configured using any IV, always encrypt with a random IV
   SymmetricEncryption.cipher = SymmetricEncryption::Cipher.new(
-      key:         ENV['VAULT_KEY'][0...32],
-      iv:          ENV['VAULT_IV'][0...16],
-      cipher_name: 'aes-256-cbc'
+    version:            1,
+    always_add_header:  true,
+    cipher_name:        'aes-256-cbc',
+    key:                hkdf.next_bytes(32),
   )
+
+  # setup legacy cipher for decoding values using the legacy truncated key + fixed IV
+  if vault_iv = ENV['VAULT_IV']
+    # legacy cipher used for backwards-compatibility with ciphertexts encrpyted using the truncated key
+    legacy_cipher = SymmetricEncryption::Cipher.new(
+      version:     0, # default, as used for encrypting existing secrets with headers
+      cipher_name: 'aes-256-cbc',
+      key:         vault_key[0...32],
+      iv:          vault_iv[0...16],
+    )
+
+    # used to lookup cipher for vault values having a header with version=0 using the legacy key
+    SymmetricEncryption.secondary_ciphers = [
+      legacy_cipher,
```

Comment:
This secondary cipher with `version=0` is a little bit silly, because it will never actually get used in any production release, assuming that this PR gets merged before then... it will only ever get used for decrypting values generated using git master at #3184, which still entrpys values using the truncated key and writes them out to mongo with a `version: 0` header.

Existing secrets from previous releases that don't have any header will use the `SymmetricEncryption.select_cipher => legacy_cipher`.

Reply:
Actually, that's not true... the [`Configuration`](https://github.com/kontena/kontena/blob/v1.4.3/server/app/models/configuration.rb#L32) model already stores encrypted values with a header, so we do really need this `version: 0` secondary cipher, also for release versions.